tests/conftest.py

```diff
@@ -89,3 +86,126 @@ def _snapshot(**kwargs):
             yield snapshot
 
     return _snapshot
+
+
+if PY2:
+    import marshal
+    from py_compile import MAGIC
+    from py_compile import wr_long
+
+    from _pytest._code.code import ExceptionInfo
+
+    def dump_code_to_file(code, file):
+        file.write(MAGIC)
+        wr_long(file, long(time.time()))  # noqa
+        marshal.dump(code, file)
+        file.flush()
+
+
+else:
+    import importlib
+
+    code_to_pyc = getattr(
+        importlib._bootstrap_external, "_code_to_bytecode" if sys.version_info < (3, 7) else "_code_to_timestamp_pyc"
+    )
+
+    def dump_code_to_file(code, file):
+        file.write(code_to_pyc(code, time.time(), len(code.co_code)))
+        file.flush()
+
+
+def unwind_params(params):
+    if params is None:
+        yield None
+        return
+
+    for _ in product(*(((k, v) for v in vs) for k, vs in params.items())):
+        yield dict(_)
+
+
+def run_function_from_file(item, params=None):
+    file, _, func = item.location
+    marker = item.get_closest_marker("subprocess")
+
+    file_index = 1
+    args = marker.kwargs.get("args", [])
+    args.insert(0, None)
+    args.insert(0, sys.executable)
+    if marker.kwargs.get("ddtrace_run", False):
+        file_index += 1
+        args.insert(0, "ddtrace-run")
+
+    env = os.environ.copy()
+    env.update(marker.kwargs.get("env", {}))
+    if params is not None:
+        env.update(params)
+
+    expected_status = marker.kwargs.get("status", 0)
+
+    expected_out = marker.kwargs.get("out", "")
+    if expected_out is not None:
+        expected_out = expected_out.encode("utf-8")
+
+    expected_err = marker.kwargs.get("err", "")
+    if expected_err is not None:
+        expected_err = expected_err.encode("utf-8")
+
+    with open(file) as f:
+        t = ast.parse(f.read())
+        for node in t.body:
+            if isinstance(node, ast.FunctionDef) and node.name == func:
```

Comment:
This does mean it has to be a top level function right?

Would it work if we had:

```python
if PY3:
    @pytest.markers.subprocess()
    def test_only_for_py3():
        pass
```

(yes, normally we'd just use a skip, but as an example)

Reply:
No, I think this wouldn't work. If we really need support for this we could define a visitor to find the `FunctionDef` instead of just looping over the module body.

Reply:
👍🏻 fine deferring, this is a complex enough change as is 😆